shared/fs/container.js

```diff
@@ -12,40 +12,71 @@ type OwnProps = {
 
 type StateProps = {
   path: Types.Path,
-  items: I.List<Types.Path>,
+  itemNames: I.List<string>,
   progress: Types.ProgressType,
+  sortSetting: Types.SortSetting,
+  username?: string,
+  pathItems: I.Map<Types.Path, Types.PathItem>,
 }
 
-type DispatchProps = {}
+type _sortSettingSetter = (setting: Types._SortSetting) => void
+type DispatchProps = {
+  loadFolderList: (path: Types.Path) => void,
+  _sortSettingChangeFactory: (path: Types.Path) => _sortSettingSetter,
+}
 
 const mapStateToProps = (state: TypedState, {routeProps}: OwnProps) => {
   const path = Types.stringToPath(routeProps.get('path', Constants.defaultPath))
+  const sortSetting = state.fs.pathUserSettings.get(path, Constants.makePathUserSetting()).get('sort')
   const itemDetail = state.fs.pathItems.get(path)
-  const items = itemDetail && itemDetail.type === 'folder' ? itemDetail.get('children', I.List()) : I.List()
+  const itemNames =
+    itemDetail && itemDetail.type === 'folder' ? itemDetail.get('children', I.List()) : I.List()
   const progress: Types.ProgressType = itemDetail ? itemDetail.progress : 'pending'
-  return {items, path, progress}
+  const username = state.config.username || undefined
+  const pathItems = state.fs.pathItems
+  return {itemNames, path, progress, sortSetting, username, pathItems}
 }
 
 const mapDispatchToProps = (dispatch: Dispatch) => ({
-  _loadFolderList: (path: Types.Path) => dispatch(FsGen.createFolderListLoad({path})),
+  loadFolderList: (path: Types.Path) => dispatch(FsGen.createFolderListLoad({path})),
+  _sortSettingChangeFactory: (path: Types.Path) => (setting: Types.SortSetting) =>
+    dispatch(FsGen.createSortSetting({path: path, sortSetting: Constants.makeSortSetting(setting)})),
 })
 
-const mergeProps = ({path, items, progress}: StateProps, dispatchProps: DispatchProps, ownProps) => ({
-  items: items.map(name => Types.pathConcat(path, Types.pathToString(name))).toArray(),
-  progress,
-  path,
-  /* TODO: enable these once we need them:
-  name: Types.getPathName(stateProps.path),
-  visibility: Types.getPathVisibility(stateProps.path),
-  */
-  ...dispatchProps,
-})
+const mergeProps = (stateProps: StateProps, dispatchProps: DispatchProps, ownProps) => {
+  const setSortSetting = dispatchProps._sortSettingChangeFactory(stateProps.path)
+  return {
+    items: Constants.sortPathItems(
+      stateProps.itemNames.map(name => Types.pathConcat(stateProps.path, name)).map(
+        p => stateProps.pathItems.get(p, Constants.makeUnknownPathItem()) // provide an unknown default to make flow happy
+      ),
+      stateProps.sortSetting,
+      Types.pathIsNonTeamTLFList(stateProps.path) ? stateProps.username : undefined
+    )
+      .map(({name}) => Types.pathConcat(stateProps.path, name))
+      .toArray(),
+    progress: stateProps.progress,
+    path: stateProps.path,
+    sortSetting: stateProps.sortSetting,
+
+    setSortByNameAsc: () => setSortSetting({sortBy: 'name', sortOrder: 'asc'}),
```

Comment:
Why do we need custom methods for each of these? Can we not just call `setSortSetting({sortBy: 'name', sortOrder: 'asc'})` in the component itself?

Reply:
I had the "toggle" version in SortBar and @chrisnojima requested to move that into the container, so I followed same pattern here.

But this is also slightly different since it doesn't really have any logic based on existing props. So I'm not sure what the best practice is here now.

Reply:
IMO this should be a `setSortSetting(sortBy, sortOrder)`, and it handles creating the object and triggering the actions. That way we only need to pass one function around, and all these `setSortBy{field}{dir}` properties can go away.

Reply:
Hmm that means we can't simply do e.g.  `<ClickableBox onClick={props.setSortByNameAsc} ... />` anymore, but it'll need to be `<ClickableBox onClick={() => props.setSortSetting('name', 'asc)} ... />`. So it's a bit more verbose inside `render()`

Reply:
IMO that's fine, unless @chrisnojima says otherwise. You save like 10 lines of code for each one of those.

Reply:
You can also make `setSortSetting` into `setSortSettingFn(sortBy, sortOrder)`, which returns a `() => void` that does the right thing.

Reply:
Oh I like the `setSortSettingFn(sortBy, sortOrder)` suggestion! @chrisnojima are you OK with that?

Reply:
Looks like I'm gonna need to refactor this a little bit in the popup thing. Will go with the function factory approach!

Reply:
i don't think you need a factory. just sort it in your connector and thats it. the actions change the sortBy/sortOrder and its just executed by the connector